source/common/config/watch_map.h

```diff
@@ -0,0 +1,121 @@
+#pragma once
+
+#include <set>
+#include <string>
+#include <utility>
+
+#include "envoy/config/subscription.h"
+
+#include "common/common/assert.h"
+#include "common/common/logger.h"
+
+#include "absl/container/flat_hash_map.h"
+#include "absl/container/flat_hash_set.h"
+
+namespace Envoy {
+namespace Config {
+
+struct Watch;
+using WatchPtr = std::unique_ptr<Watch>;
+
+struct AddedRemoved {
+  AddedRemoved(std::set<std::string> added, std::set<std::string> removed)
+      : added_(added), removed_(removed) {}
+  std::set<std::string> added_;
+  std::set<std::string> removed_;
+};
+
+// Manages "watches" of xDS resources. Several xDS callers might ask for a subscription to the same
+// resource name "X". The xDS machinery must return to each their very own subscription to X.
+// The xDS machinery's "watch" concept accomplishes that, while avoiding parallel redundant xDS
+// requests for X. Each of those subscriptions is viewed as a "watch" on X, while behind the scenes
+// there is just a single real subscription to that resource name.
+//
+// This class maintains the watches<-->subscription mapping: it
+// 1) delivers updates to all interested watches, and
+// 2) tracks which resource names should be {added to,removed from} the subscription when the
+//    {first,last} watch on a resource name is {added,removed}.
+//
+// #1 is accomplished by WatchMap's implementation of the SubscriptionCallbacks interface.
+// This interface allows the xDS client to just throw each xDS update message it receives directly
+// into WatchMap::onConfigUpdate, rather than having to track the various watches' callbacks.
+//
+// The information for #2 is returned by updateWatchInterest(); the caller should use it to
+// update the subscription accordingly.
+//
+// A WatchMap is assumed to be dedicated to a single type_url type of resource (EDS, CDS, etc).
+class WatchMap : public SubscriptionCallbacks, public Logger::Loggable<Logger::Id::config> {
+public:
+  WatchMap() {}
+
+  // Adds 'callbacks' to the WatchMap, with no resource names being watched.
+  // (Use updateWatchInterest() to add some names).
+  // Returns the newly added watch, to be used for updateWatchInterest. Destroy to remove from map.
+  WatchPtr addWatch(SubscriptionCallbacks& callbacks);
+
+  // Updates the set of resource names that the given watch should watch.
+  // Returns any resource name additions/removals that are unique across all watches. That is:
+  // 1) if 'resources' contains X and no other watch cares about X, X will be in added_.
+  // 2) if 'resources' does not contain Y, and this watch was the only one that cared about Y,
+  //    Y will be in removed_.
+  AddedRemoved updateWatchInterest(Watch* watch,
```

Comment:
It cannot. The point of WatchMap, and the reason why a standalone Watch doesn't have much meaningful logic it can take on, is that the complexity with the whole "watches" concept arises from the union of all resource names being watched. WatchMap exists to do the watch logic that needs to be looking at all watches at once - which turns out to be basically all of it.

More concretely, notice that wildcard_watches_ needs to be modified here, and if you dive down into findAdditions/Removals, you'll see those two are entirely about modifying the overall watch_interest_ map.

Reply:
Well, I'm suggesting that `Watch` has a `parent_` field pointing back to the `WatchMap` and does what `updateWatchInterest` does today; we're just replacing the `Watch * watch` argument here with a `this`.

Reply:
I see. I would prefer not to, since it adds that parent pointer, plus this new passthrough function.

I think I see where you're coming from, though; you want calling code to operate directly on the Watch rather than dealing with the WatchMap. But, this function is ultimately asking the map to do something with a particular one of its keys, which are the Watch pointers (here is where the (unfortunately less efficient) integer tokens would be clearer). And, this function is something that has meaning across all the watches, so I think avoiding interaction with the WatchMap here would actually be less clear.

Reply:
Done.